File: 304/range-sum-query-2d-immutable.js
/**
 * @param {number[][]} matrix
 */
var NumMatrix = function (matrix) {
    this.sum = [];
    for (let i = 0, a = matrix.length; i < a; i++) {
        this.sum[i] = [];
        for (let j = 0, b = matrix[i].length; j < b; j++) {
            let ans = 0;
            ans += i ? this.sum[i - 1][j] : 0;
            ans += j ? this.sum[i][j - 1] : 0;
            ans -= (i && j) ? this.sum[i - 1][j - 1] : 0;
            ans += matrix[i][j];
            this.sum[i][j] = ans;
        }
    }
};

/** 
 * @param {number} row1 
 * @param {number} col1 
 * @param {number} row2 
 * @param {number} col2
 * @return {number}
 */
NumMatrix.prototype.sumRegion = function (row1, col1, row2, col2) {
    var ans = 0;
    ans += this.sum[row2][col2];
    ans -= row1 ? this.sum[row1 - 1][col2] : 0;
    ans -= col1 ? this.sum[row2][col1 - 1] : 0;
    ans += (row1 && col1) ? this.sum[row1 - 1][col1 - 1] : 0;
    return ans;
};

/** 
 * Your NumMatrix object will be instantiated and called as such:
 * var obj = Object.create(NumMatrix).createNew(matrix)
 * var param_1 = obj.sumRegion(row1,col1,row2,col2)
 */

File: 304/range-sum-query-2d-immutable.go
type NumMatrix struct {
	dp [][]int
}

func Constructor(matrix [][]int) NumMatrix {
	var m, n int
	if len(matrix) == 0 || len(matrix) == 0 {
		m, n = 0, 0
	} else {
		m, n = len(matrix), len(matrix[0])
	}

	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			dp[i][j] = matrix[i-1][j-1] + dp[i-1][j] + dp[i][j-1] - dp[i-1][j-1]
		}
	}

	return NumMatrix{dp: dp}
}

func (this *NumMatrix) SumRegion(row1 int, col1 int, row2 int, col2 int) int {
	return this.dp[row2+1][col2+1] - this.dp[row2+1][col1] - this.dp[row1][col2+1] + this.dp[row1][col1]
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * obj := Constructor(matrix);
 * param_1 := obj.SumRegion(row1,col1,row2,col2);
 */